src/common/device/AndroidDevice.ts

```diff
@@ -103,18 +107,95 @@ export class AndroidDevice implements BaseDevice {
+     * @param launchArguments Extra arguments to be passed to the app upon launch.
      */
-    public async launchApp(
-        target: string,
-        appBundlePath?: string,
-        targetAppArguments?: LaunchArgument[]
-    ): Promise<void> {
-        await AndroidUtils.launchAppInBootedEmulator(this.port, target, appBundlePath, targetAppArguments, this.logger);
+    public async launchApp(target: string, appBundlePath?: string, launchArguments?: LaunchArgument[]): Promise<void> {
+        await AndroidUtils.launchAppInBootedEmulator(this.port, target, appBundlePath, launchArguments, this.logger);
+    }
+
+    /**
+     * Checks to see if a certificate is already installed on the device.
+     *
+     * @param certData An SSLCertificateData object containing the certificate data.
+     * @returns A boolean indicating if a certificate is already installed on the device or not.
+     */
+    public async isCertInstalled(certData: SSLCertificateData): Promise<boolean> {
+        // For Android emulators, certificates are installed in files named as their subject hash.
+        // We can query the device to see if it has the file with the name as the subject hash of the
+        // provided certificate. If so then we can say that the cert is already installed.
+        const subjectHash = CryptoUtils.getSubjectHashOld(certData);
+        const fileName = `${subjectHash}.0`; // this is the special file name
+
+        try {
+            // start adb as root so that we can query specific folder location
+            await AndroidUtils.executeAdbCommand('root', this.port, this.logger);
+
+            // see if the file exists already
+            const result = await AndroidUtils.executeAdbCommand(
+                'shell "ls /data/misc/user/0/cacerts-added"',
+                this.port,
+                this.logger
+            );
+            return result.includes(fileName);
+        } catch (error) {
+            // If file doesn't exist then an error is thrown on some versions of adb.
+            // Also determining if cert is already installed is a best effort so if any errors
+            // occur we will continue and assume that cert is not installed.
+            this.logger?.warn(error);
+        }
+        return false;
+    }
+
+    /**
+     * Installs a certificate on the device.
+     *
+     * @param certData An SSLCertificateData object containing the certificate data.
+     */
+    public async installCert(certData: SSLCertificateData): Promise<void> {
+        // For Android emulators, we need to save the PEM certificate in a file named as the cert subject hash.
+        const subjectHash = CryptoUtils.getSubjectHashOld(certData);
+        const pemContent = certData.pemCertificate ?? CryptoUtils.derToPem(certData.derCertificate);
+        const fileName = `${subjectHash}.0`; // this is the special file name
+        const certFilePath = path.join(os.tmpdir(), fileName);
```

Comment:
Might be better to make use of `CommonUtils.createTempDirectory()` instead of `os.tmpdir()` here, as that's typically how we've standardized creating temporary files/folders in the project.

Reply:
With that approach we will have a unique directory created each time `CommonUtils.createTempDirectory()` is called. I would then need to write extra code logic around deleting that directory and the cert file in it, after cert is installed (or an error occurs).

But using `os.tmpdir()` directly means that the certfile (which is always named the same) will always be created in the same location (overwriting if it already exists) and so we don't need cleanup logic (the file is less than 1kb and is reused/overwritten in subsequent calls).